### ISSUE TITLE
Sprite generation from TTF loaded via AssetManager is inconsistent across devices

### ISSUE THREAD

**Issue by vlaz** [libgdx/libgdx#2859]
This is my first post here, I apologize in advance if it stinks.

I'm working on a resolution-independent game whereby I load graphics as vectors using TTFs and convert them to textures/sprites based on device requirements.  This was working flawlessly and consistently across all tested devices until I employed the AssetManager to load the fonts.  Now, glyph data appears to vary across devices and produces unexpected sprites.  Note, I've had no issues on desktop and drawing directly from the bitmap font is still working correctly.  However, I require rotation and tinting so as far as I know, I'm stuck with using sprites.

I've provided all relevant code snippets below just to make sure I'm not doing anything incorrectly, as well as screenshots of the inconsistency.

AssetLoader.load()
```java
am								= new AssetManager();
FileHandleResolver resolver		= new InternalFileHandleResolver();
am.setLoader(FreeTypeFontGenerator.class, new FreeTypeFontGeneratorLoader(resolver));
am.setLoader(BitmapFont.class, ".ttf", new FreetypeFontLoader(resolver));

FreeTypeFontLoaderParameter p	= new FreeTypeFontLoaderParameter();
p.fontParameters.size			= (int) (Gdx.graphics.getHeight() / 10);
p.fontParameters.flip			= true;
p.fontParameters.minFilter		= TextureFilter.Linear;
p.fontParameters.characters		= "PY023";
									// P down feather
									// Y quill feather
									// 0 full star, no outline
									// 2 half star and outline
									// 3 empty star, just outline
p.fontFileName					= "f.ttf";
am.load("f.ttf", BitmapFont.class, p);
```

AssetLoader.loadProgress()
```java
if(am.update()) {
	f		= am.get("f.ttf");
	fQuill	= Utils.getGlyphSprite('Y', f);
	fDown	= Utils.getGlyphSprite('P', f);
	sFull	= Utils.getGlyphSprite('0', f);
	sHalf	= Utils.getGlyphSprite('2', f);
	sEmpty	= Utils.getGlyphSprite('3', f);
}
return am.getProgress();
```

Utils.getGlyphSprite(char ch, BitmapFont f)
```java
Glyph g		= f.getData().getGlyph(ch);
Sprite s	= new Sprite(f.getRegion().getTexture(), g.srcX, g.srcY, g.width, g.height);
s.flip(false, true);
s.setOrigin(g.width / 2, g.height / 2);
return s;
```

As an example, here is the painting of just the star outlines.

GameRenderer.paint()
```java
spritebatch.draw(empty, loc.x - empty.getWidth() / 2, loc.y - empty.getHeight() / 2, empty.getOriginX(), empty.getOriginY(), empty.getWidth(), empty.getHeight(), scale.current, scale.current, rotation.current + 90);
```

This is the expected result, observed on desktop and a Nexus 4, Android 5.0.1.
![screenshot_2015-02-18-17-27-30clip](https://cloud.githubusercontent.com/assets/11069270/6262172/0ad54e30-b7cc-11e4-8f93-a9adf8f0f1a4.png)

This is a garbled result, observed on an HTC One S, Android 4.1.1; note the feathers in place of stars, though they're not correctly drawn either, the framing seems to be off a bit.
![screenshot_2015-02-18-22-38-27clip](https://cloud.githubusercontent.com/assets/11069270/6262175/0f7bddaa-b7cc-11e4-884b-d9066069ef1f.png)

The Nexus 4 isn't entirely perfect either, it had garbled results too via the AssetManager until I started loading the font in sections.  Desktop has always been perfect though, which is why I think this is more likely a bug in LibGDX than in my code.

Please let me know if there is any other information I can provide.

**Comment by MobiDevelop** [comment]
So, it works correctly if you use FreetypeFontGenerator directly, but not when you use FreeTypeFontLoader?

**Comment by vlaz** [comment]
Good catch, maybe deprecate it?  Most programmers are a bit OCD and wouldn't mind the change.

And, that's correct.  It may also be of some importance to note that the feather sprites look fine.  So going by the character selection list:

ch|icon|status|result
---|---|---|---
P|down feather|good|
Y|quill feather|good|
0|full star, no outline|good|
2|half star and outline|bad|P but offset
3|empty star, just outline|bad|Y but offset

**Comment by rinold** [comment]
BTW, why are you flipping font, and then flipping sprite - does your issue reproduces without flip?

Also creating the sprite with integer x, y, w, h as you do sets region to:
```java
    setRegion(x * invTexWidth, y * invTexHeight, (x + width) * invTexWidth, (y + height) * invTexHeight);
```
while for the Glyph the [more complicated logic](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/graphics/g2d/BitmapFont.java#L209) seems to be used to calculate the u,v values including the flipping, x and y offset values. If you  try to create the sprite without flip and use direct Glyph u,v values like this:
```
... = new Sprite(f.getRegion().getTexture(), g.u, g.v g.u2, g.v2);
```
does your issue still reproduces?

**Comment by vlaz** [comment]
Thanks for your input!  It seems that the second flip doesn't undo the first.  When the sprite is created, the first flip is ignored, so I do need the second one.  Either way, removing it does not correct the issue.

Your point on the glyph 'word wrap' makes sense.  When I increase the size of the font at creation time, I can replicate the issue in the Nexus as well.  However, using u/v is not straightforward.  There is no constructor for floats.  I attempted to multiply by width and height and cast to int but got the same garbled results.

**Comment by vlaz** [comment]
I did some more experimentation and figured it out...

I have to retract the claim that this only occurs with AssetManager.  I was able to replicate the issue with the usual loading method.  I can attest that the problem was not evident previous to using AssetManager.  I played around with it before and it seemed to be the only factor, but evidently there must have been something else I missed.  Sorry for the confusion.

I tried using a very reliable font to make sure it wasn't something in my TTF and confirmed that it wasn't that.

Lastly, I figured out how to get the u/v coordinates to work and after plotting with them instead, the issue was still there, which made sense because plotting the coordinates, you can tell that the dimensions overlap.

Then it hit me that this may be producing several textures and turns out Glyph has a page index and Font has a getRegions() method.  I constructed the sprite by querying the glyph's page:

```java
Sprite s = new Sprite(f.getRegion(g.page).getTexture(), g.srcX, g.srcY, g.width, g.height);
```

and now it works at any size!

So this can be closed as it's not a bug.  Thanks for your help.

Just as a discussion point, I understand getRegion() is there for back-compatibility, but perhaps it should complain when the font has multiple pages now that that's possible.

**Comment by rinold** [comment]
Your problems is that by default the max texture size is 1024, and the resulting generated glyphs are big and as result - splitted on pages. Looking at the table you provided and at your screenshots, I suggest you to play with glyph sizes (downsize them) - as they seems to be too huge for your real usage - do you really need symbols with 512 sizes while on screenshots they are looking much smaller? (may be I'm mistaken but it seems so for me). If you scale their size down, it will not only consume less memory, but will remove unnecessary texture switching between pages while drawing them.

**Comment by vlaz** [comment]
Yes, i definitely don't need that size, I was just trying to replicate the issue on my Nexus and Desktop environments.  On the HTC, the issue occurred at 96 px (screen height / 10).  As it could only fit 3 glyphs on the first page, it must have a much smaller default texture size.
